### ISSUE TITLE
Host based routing

### ISSUE THREAD

**Issue by airhorns** [fastify/fastify#2498]
Is there a way to restrict some fastify handlers to only match on certain request hosts? Similar to the Rails router's `constraints: { host: "somehost.com" }` piece that lets you scope certain routes to only match if certain conditions met.

My use case is the classic one-subdomain-per-customer setup where I have a wildcard DNS setup that points `*.airhorns.dev` to my Fastify instance , so that customer Foo's app runs at `foo.airhorns.dev` and customer Bar's app runs at `bar.airhorns.dev`. I have this all set up and working, but, I want to reserve `auth.airhorns.dev` and `app.airhorns.dev` for particular platform level functionality for login and whatnot. I need different route handlers to run for those domains vs the normal ones that might serve a customer's content. After trying a few things and poking around the router code I'm not sure it's possible. I also really don't want every route handler to have to check which host it is running on and do one of two completely different things depending -- that seems like a huge smell. 

Is it?

Thanks!

**Comment by jsumners** [comment]
Route the domains to different servers and launch the servers with configuration that indicates which domain it should be processing (e.g. via an environment variable). See https://github.com/fastify/fastify/blob/7ac9da63f7d0ab03dd0548fdb52db03d3295f3e7/docs/Recommendations.md#reverseproxy

**Comment by RafaelGSS** [comment]
Like the document above says. Why you don't do this at reverse proxy?

**Comment by airhorns** [comment]
FWIW I already have a reverse proxy as well and agree with many of the justifications in that document -- I use it for SSL termination and load balancing and no downtime deploys and all that other goodness.

At a minimum though I want my server instance to handle requests for more than one domain because the domains aren't known in advance. I can't pass in all the subdomains via an environment variable or what have you because they're in the database and frequently changing as new tenants sign up or churn or what have you. An unrecognized subdomain needs to be 404'd and a recognized one needs to be rendered for that particular tenant, so if you ask me, I'm already doing host based routing, it's just in the route handler.

That said, these route handlers are still "homogenous" in that `foo.airhorns.dev/handler` runs the same code as `bar.airhorns.dev/handler`, and I think maybe what you're saying is that if `foo.airhorns.dev/handler` and `auth.airhorns.dev/handler` need to run different handlers because the auth host is special, I need to run two different fastify instances to accomplish this. 

I think this is kind of weird for a number of reasons:

 - the complexity in this setup exists and it has to go somewhere. Pushing into the load balancing tier actually makes my job as a developer and operator harder -- now I have to mirror that configuration in development, test that it agrees between the load balancer and the fastify instances, and care about a whole other router that is doing _stuff_ with requests when I already have a router in fastify handy. The fastify router is already handling super important routing for my whole application, has a facility for testing it, and is the natural place a developer on my team would look to make changes to how the routing works, why split it up across two systems if one would do?
 - this setup would be completely possible in Fastify if I were using path segments instead, like `airhorns.dev/foo` and `airhorns.dev/auth`, so why is host based routing somehow different? I understand that it's a big body of work to accomplish this and I totally get if it's not really a priority -- that's a fine answer, but it's a different answer than "we actually don't want to support this in Fastify because it's the wrong pattern"
 - Rails definitely supports this fine so I think its totally possible to write a performant router that can route using the host, notably express doesn't though unless you start using multiple routers as middlewares
 - There's an operational cost to running one service per domain. You pay more for all the memory the extra instances consume, you get worse elasticity because now you have dedicated pools with different scaling, you have to care about which logs stream you're looking at, you start having to debug across instances that are talking to each other, you have to build multiple containers instead of just one, etc etc. I'd call all of these things minor inconveniences, but again, why mandate the developer manage that inconvenience instead of solving it at the framework level?

I guess what I am asking for is framework support for a very real use case instead of having to resort to configuring the same process to do totally different things with env variables or being forced prematurely to go to a services based architecture for a small concern like the router's feature set.

**Comment by mcollina** [comment]
Those are all valid motives. I would say no one of us had the need to implement this - or maybe the cost of development time vs possible savings was too steep to justify.

Note that what you need it's not pretty but it's 100% achievable by having multiple fastify instances on the same Node.js process. This will have some added memory cost, but likely smaller than running multiple processes. Check out the [`serverFactory`](https://www.fastify.io/docs/latest/Server/#serverfactory) option.

**Comment by airhorns** [comment]
Oh, I always thought this was a pretty common pattern? Shopify, Webflow, Vercel, Netlify, Heroku, Slack, etc etc all provision one subdomain per customer like this, and some have their own corporate stuff running on the same TLD as their customer stuff. 

Out of curiosity, has anyone thought through what a good API for this might be? Is that part of why it's been a wontfix for now is that it's difficult to figure out a way to add it into the current Fastify route registration API?

**Comment by jsumners** [comment]
It's not a "wontfix". It's a feature request. Regardless, issues are resolved primarily by those who are interested in seeing the issue addressed. Given that no one on the core team has experienced a need for this feature, it is unlikely anyone on that core team will take time to address it. But we are very willing to review PRs for it.

If it's not something you have the time to work on yourself, there are consultancies that will gladly do the work on your behalf.

**Comment by airhorns** [comment]
> It's not a "wontfix"

My mistake! I thought thats what you meant with the reverse proxy solution, my bad.

**Comment by mcollina** [comment]
> Oh, I always thought this was a pretty common pattern? Shopify, Webflow, Vercel, Netlify, Heroku, Slack, etc etc all provision one subdomain per customer like this, and some have their own corporate stuff running on the same TLD as their customer stuff.

They do not serve their whole system through the same Node.js (or any other language) process. They use load balancer and infra to handle this case.

**Comment by airhorns** [comment]
I actually used to work on Shopify and Shopify does it with the Rails router, but I don't know about the rest of the internal implementations. It made lots of things easier to have the different domains served by the same process at Shopify especially as users were being redirected between the different domains for things like session bootstrapping. If you're forwarding someone from a global auth area to a specific subdomain on a different apex domain (like app.shopify.com -> airhorns.myshopify.com) and you need to keep their login state by setting up a cookie on the new domain, you can have the same code create the cross domain request as the code that processes the cross domain request. The same code shares the same config for the secrets and cookie signing logic etc, so if you want to say rotate your cookie secret or whatever (like `fastify-secure-session` wonderfully supports) you can just ship that code once instead of having to coordinate that across many different processes. I know that code and config sharing between projects is possible, but again I am just trying to make the point that in an ideal, everyone has infinite time to work on everything world, the internal feature set of the framework's router wouldn't dictate the architecture of the whole app. I realize we don't live in that world.

So, how would you folks tackle this issue if someone wanted to? I feel like it's kind of a doozy for a newbie to Fastify core to work on but maybe if we can establish a path forward I can find some time or money to get it done!

**Comment by mcollina** [comment]
I think the starting point must be a PR to find-my-way, our router.

**Comment by delvedor** [comment]
Very likely the most performant solution to this problem is a similar approach that we use for versioned routes. We need a fast algorithm that can handle the host header and compre it with the host configuration passed by the user.
Once we find a good algorithm for doing this, the integration should be easy enough :)

**Comment by airhorns** [comment]
I am happy to take a stab at that PR but before diving in I wonder what the right developer-facing interface would be for restricting certain routes to certain hosts. Would you folks imagine it in another option passed to `.route` or as part of the `url` parameter passed to the route shorthand methods? I feel like it might be kind of tricky to make the URL parsing logic completely unambiguous such that host-ful route patterns and host-less route patterns are really clearly one or the other. 

I can shed a little light on the issue with my experience with the Rails router. The way the Rails router does this is with a full blown route option called `constraints` that lets you match on any aspect of the incoming route however you wish. You can do `constraints: { host: "auth.airhorns.dev" }` or `{constraints: { host: /[^.]+\.airhorns.dev/}`, or run full on code with `constraints: SomeFancyConstraintClass.new` (see https://guides.rubyonrails.org/routing.html#advanced-constraints) .  Is that something you would be interested in in order to allow code to run during the routing phase to test if a route matched? That's the "general" solution to this problem if you ask me and would mean it'd be perma-fixed for any other strange setups. It'd have to be developer beware in terms of performance if you're evaluating your own code to test every route's matchingness but I think there's no real way around that other than just making it less powerful. 

Also notably the Rails router uses a finite state automaton that it interprets for each request to match the route. I don't think that's necessary for Fastify and that the radix tree + associated matchers would do but I'm not really certain. Rails' routes let you add arbitrary constrained contexts that the routes defined inside must meet which would add "nodes" to the automaton's journey that couldn't actually be matched as a route but would have to be passed through, but in fastify's case, I don't think there are those same routing contexts. Plugins can't restrict all the routes defined in their scope somehow right?

**Comment by mcollina** [comment]
> Is that something you would be interested in in order to allow code to run during the routing phase to test if a route matched?

no. We want the route to be fully decided by the router itself. This is a key principle in find-my-way.
As @delvedor said, the approach that best match what would be needed is the one we use for versions. See how that is implemented.

**Comment by airhorns** [comment]
> This is a key principle in find-my-way.

What's the intention with this principle? Just to keep performance high?

**Comment by mcollina** [comment]
> What's the intention with this principle? Just to keep performance high?

Yes, exactly.

**Comment by airhorns** [comment]
Is anyone aware of any other feature requests similar to this one but for other route constraints? The `find-my-way` code currently treats versions in a really special way, and it feels like adding host based routing is going to make it significantly more complicated and hard to follow. If there was a 3rd or a 4th special thing like this it might make sense to refactor all of the special things to use a similar flow.

**Comment by mcollina** [comment]
If there is a generic way that does not slow things down, go for it.
